File: swift/FeedbackService.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

// MARK: - FeedbackService
// Sends feedback via Telegram, Discord, Slack, or Proxy depending on FeedbackConfig.channel.
// Uses native URLSession with async/await -- no third-party dependencies.

enum FeedbackError: LocalizedError {
    case emptyFeedback
    case invalidURL
    case serverError(String)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .emptyFeedback:
            return "Feedback text cannot be empty."
        case .invalidURL:
            return "Invalid API URL."
        case .serverError(let message):
            return "API error: \(message)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}

enum FeedbackService {

    /// Sends feedback using the configured delivery channel.
    /// - Parameter category: optional topic (e.g. "Bug"). Pass `nil` to omit.
    static func send(text: String, category: String? = nil, session: URLSession = .shared) async throws {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw FeedbackError.emptyFeedback }

        let request: URLRequest
        switch FeedbackConfig.channel {
        case .proxy:
            request = try buildProxyRequest(text: trimmed, category: category)
        default:
            let message = buildMessage(text: trimmed, category: category)
            switch FeedbackConfig.channel {
            case .telegram:
                request = try buildTelegramRequest(message: message)
            case .discord:
                request = try buildDiscordRequest(message: message)
            case .slack:
                request = try buildSlackRequest(message: message)
            case .proxy:
                fatalError("Unreachable")
            }
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw FeedbackError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw FeedbackError.serverError("Invalid response")
        }

        // Discord returns 204 No Content on success
        let isSuccess = httpResponse.statusCode >= 200 && httpResponse.statusCode < 300
        if !isSuccess {
            let responseBody = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw FeedbackError.serverError(responseBody)
        }
    }

    // MARK: - Request Builders

    private static func buildTelegramRequest(message: String) throws -> URLRequest {
        let token = FeedbackConfig.botToken
        guard let url = URL(string: "https://api.telegram.org/bot\(token)/sendMessage") else {
            throw FeedbackError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "chat_id": FeedbackConfig.chatID,
            "text": message,
            "parse_mode": "Markdown"
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return request
    }

    private static func buildDiscordRequest(message: String) throws -> URLRequest {
        let webhookURL = FeedbackConfig.webhookURL
        guard let url = URL(string: webhookURL) else {
            throw FeedbackError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["content": message]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return request
    }

    private static func buildSlackRequest(message: String) throws -> URLRequest {
        let webhookURL = FeedbackConfig.webhookURL
        guard let url = URL(string: webhookURL) else {
            throw FeedbackError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["text": message]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return request
    }

    private static func buildProxyRequest(text: String, category: String?) throws -> URLRequest {
        guard let url = URL(string: FeedbackConfig.proxyURL) else {
            throw FeedbackError.invalidURL
        }

        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "Unknown"

        #if os(iOS)
        let platform = "iOS \(UIDevice.current.systemVersion)"
        #else
        let platform = ProcessInfo.processInfo.operatingSystemVersionString
        #endif

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: Any] = [
            "app_name": FeedbackConfig.appName,
            "app_version": appVersion,
            "platform": platform,
            "feedback": text
        ]
        if let category, !category.trimmingCharacters(in: .whitespaces).isEmpty {
            body["category"] = category
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return request
    }

    // MARK: - Message Formatting

    static func buildMessage(text: String, category: String? = nil) -> String {
        let appName = FeedbackConfig.appName
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "Unknown"
        let osVersion = ProcessInfo.processInfo.operatingSystemVersionString

        #if os(iOS)
        let platformEmoji = "\u{1F34E}"
        let platform = "iOS \(UIDevice.current.systemVersion)"
        #else
        let platformEmoji = "\u{1F4F1}"
        let platform = osVersion
        #endif

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm"
        formatter.timeZone = TimeZone(identifier: "UTC")
        let timestamp = formatter.string(from: Date())

        var lines = [
            "\u{1F4F1} App: \(appName)",
            "\u{1F4E6} Version: \(appVersion)",
            "\(platformEmoji) Platform: \(platform)",
            "\u{1F554} Time: \(timestamp) UTC"
        ]

        if let category, !category.trimmingCharacters(in: .whitespaces).isEmpty {
            lines.append("\u{1F3F7}\u{FE0F} Type: \(category)")
        }

        lines.append(contentsOf: [
            "",
            "\u{1F4AC} Feedback:",
            text
        ])

        return lines.joined(separator: "\n")
    }
}

File: kotlin/FeedbackService.kt
package com.feedpush

import android.content.Context
import android.os.Build
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

// MARK: FeedbackService
// Sends feedback via Telegram, Discord, or Slack depending on FeedbackConfig.channel.
// Uses native HttpURLConnection with coroutines -- no third-party dependencies.

sealed class FeedbackResult {
    data object Success : FeedbackResult()
    data class Error(val message: String) : FeedbackResult()
}

object FeedbackService {

    suspend fun send(context: Context, text: String): FeedbackResult {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return FeedbackResult.Error("Feedback text cannot be empty.")

        val message = buildMessage(context, trimmed)

        return withContext(Dispatchers.IO) {
            try {
                val (url, body) = when (FeedbackConfig.channel) {
                    FeedbackChannel.TELEGRAM -> buildTelegramPayload(message)
                    FeedbackChannel.DISCORD -> buildDiscordPayload(message)
                    FeedbackChannel.SLACK -> buildSlackPayload(message)
                }

                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                connection.connectTimeout = 15_000
                connection.readTimeout = 15_000

                connection.outputStream.use { os ->
                    os.write(body.toString().toByteArray(Charsets.UTF_8))
                }

                val responseCode = connection.responseCode

                // Discord returns 204 No Content on success
                if (responseCode in 200..299) {
                    FeedbackResult.Success
                } else {
                    val errorBody = connection.errorStream?.let { stream ->
                        BufferedReader(InputStreamReader(stream)).use { it.readText() }
                    } ?: "Unknown error"
                    FeedbackResult.Error("API error ($responseCode): $errorBody")
                }
            } catch (e: Exception) {
                FeedbackResult.Error("Network error: ${e.localizedMessage ?: e.message ?: "Unknown error"}")
            }
        }
    }

    // MARK: Payload Builders

    private fun buildTelegramPayload(message: String): Pair<URL, JSONObject> {
        val token = FeedbackConfig.botToken()
        val url = URL("https://api.telegram.org/bot$token/sendMessage")
        val body = JSONObject().apply {
            put("chat_id", FeedbackConfig.chatID)
            put("text", message)
            put("parse_mode", "Markdown")
        }
        return url to body
    }

    private fun buildDiscordPayload(message: String): Pair<URL, JSONObject> {
        val url = URL(FeedbackConfig.webhookURL())
        val body = JSONObject().apply {
            put("content", message)
        }
        return url to body
    }

    private fun buildSlackPayload(message: String): Pair<URL, JSONObject> {
        val url = URL(FeedbackConfig.webhookURL())
        val body = JSONObject().apply {
            put("text", message)
        }
        return url to body
    }

    // MARK: Message Formatting

    internal fun buildMessage(context: Context?, text: String): String {
        val appName = FeedbackConfig.appName

        val appVersion = try {
            context?.packageManager?.getPackageInfo(context.packageName, 0)?.versionName ?: "Unknown"
        } catch (_: Exception) {
            "Unknown"
        }

        val platform = "Android ${Build.VERSION.RELEASE}"

        val formatter = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        val timestamp = formatter.format(Date())

        return buildString {
            appendLine("\uD83D\uDCF1 App: $appName")
            appendLine("\uD83D\uDCE6 Version: $appVersion")
            appendLine("\uD83E\uDD16 Platform: $platform")
            appendLine("\uD83D\uDD54 Time: $timestamp UTC")
            appendLine()
            appendLine("\uD83D\uDCAC Feedback:")
            append(text)
        }
    }
}
